### PR TITLE
vstart: Now all OSDs are starting in parallel. Use --no-parallel to revert to sequential.

### PR DESCRIPTION
**Pull request by aclamk**
Adds option "--no-parallel". 
By default all OSDs are started in parallel, this is useful when performing lengthy startup process, like when running fsck.
With "--no-parallel" vstart reverts to sequential starting of OSDs.

### CODE REVIEW

File: src/vstart.sh [hunk 1]
```diff
@@ -153,6 +153,7 @@ smallmds=0
 short=0
 ec=0
 ssh=0
+parallel=false
```
Comment (tchaikov) [inline src/vstart.sh]: i am wondering if we should make it "true" by default? as it's always nice to have a running cluster sooner than later.

File: src/vstart.sh [hunk 2]
```diff
@@ -877,10 +883,23 @@ EOF
...
             extra_seastar_args="--smp 1 --cpuset $osd"
         fi
+        local osd_pid
         run 'osd' $osd $SUDO $CEPH_BIN/$ceph_osd \
             $extra_seastar_args $extra_osd_args \
-            -i $osd $ARGS $COSD_ARGS
+            -i $osd $ARGS $COSD_ARGS &
+        osd_pid=$!
+        if $parallel
+        then
```
Comment (tchaikov) [inline src/vstart.sh]: i'd suggest put
```sh
if $parallel; then
```
more compact this way. and it also follows the same style in this source file.

File: src/vstart.sh [hunk 3]
```diff
@@ -877,10 +883,23 @@ EOF
...
+            osds_wait=$osd_pid
+        else
+            wait $osd_pid
+        fi
     done
+    if $parallel
+    then
```
Comment (tchaikov) [inline src/vstart.sh]: ditto.

Comment (aclamk) [reply src/vstart.sh]: @tchaikov DNM because I have to fix exiting when some error occurs.

**Comment by sebastian-philipp** [comment]
jenkins test dashboard backend

**Comment by sebastian-philipp** [comment]
(^^ makes use of vstart)

**Comment by sebastian-philipp** [comment]
```
2019-11-20T12:01:09.537+0000 7febbb881700  4 mgr[dashboard] [CephFS] Removing directory: /pictures
2019-11-20T12:01:09.537+0000 7febbb881700  1 -- 127.0.0.1:0/815954651 --&gt; [v2:127.0.0.1:6834/1129835784,v1:127.0.0.1:6835/1129835784] -- client_request(unknown.0:3 rmdir #0x1/pictures 2019-11-20T12:01:09.538065+0000 caller_uid=1001, caller_gid=1001{}) v4 -- 0x6e25600 con 0x6dc5f80
2019-11-20T12:01:09.538+0000 7febaeea8700  1 -- 127.0.0.1:0/815954651 &lt;== mds.0 v2:127.0.0.1:6834/1129835784 5 ==== client_reply(???:3 = -39 (39) Directory not empty safe) v1 ==== 27+0+0 (crc 0 0 0) 0x6e25600 con 0x6dc5f80
2019-11-20T12:01:09.542+0000 7febbb881700  0 mgr[dashboard] [20/Nov/2019:12:01:09] HTTP Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/cherrypy/_cprequest.py", line 656, in respond
    response.body = self.handler()
  File "/usr/lib/python2.7/site-packages/cherrypy/lib/encoding.py", line 188, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cptools.py", line 221, in wrap
    return self.newhandler(innerfunc, *args, **kwargs)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/dashboard/services/exception.py", line 88, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/cherrypy/_cpdispatch.py", line 34, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/dashboard/controllers/__init__.py", line 657, in inner
    ret = func(*args, **kwargs)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/dashboard/controllers/__init__.py", line 850, in wrapper
    return func(*vpath, **params)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/dashboard/controllers/cephfs.py", line 375, in rm_dir
    cfs.rm_dir(path)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/src/pybind/mgr/dashboard/services/cephfs.py", line 127, in rm_dir
    self.cfs.rmdir(path)
  File "cephfs.pyx", line 923, in cephfs.LibCephFS.rmdir (cephfs.c:9856)
ObjectNotEmpty: error in rmdir /pictures: Directory not empty [Errno 39]
```

No idea if this is related.

**Comment by tchaikov** [comment]
i don't think it's related.

File: src/vstart.sh [hunk 4]
```diff
@@ -877,10 +883,23 @@ EOF
...
+        if $parallel; then
+            osds_wait=$osd_pid
+        else
+            wait $osd_pid
+        fi
     done
+    if $parallel; then
+        for p in $osds_wait; do
+            wait $p
+        done
```
Comment (batrick) [inline src/vstart.sh]: you can simply use `wait` and avoid recording the `osd_pid`s and this loop.

**Comment by ideepika** [comment]
[wip-deepika-testing](https://trello.com/c/NbhkIeBl/773-wip-deepika-testing)

File: src/vstart.sh (hunk 4 above)
Comment (tchaikov) [reply src/vstart.sh]: @aclamk ping?

Comment (aclamk) [reply src/vstart.sh]: @batrick We must have loop. `wait` with single pid copies return value from process. `wait` without pid waits for everyting to complete and succeds.

Comment (batrick) [reply src/vstart.sh]: > `wait` without pid waits for everyting to complete and succeds.

Isn't that what you want to do here?

Comment (aclamk) [reply src/vstart.sh]: No, when I run multiple processes in background, and some of them have non-zero return, then `wait` is ok with this, while with `wait <pid>` I got error.

Comment (batrick) [reply src/vstart.sh]: got it

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-12-02_17:51:44-rados-wip-kefu-testing-2019-12-02-2333-distro-basic-mira/

failures tracked by

- https://tracker.ceph.com/issues/43132
- https://tracker.ceph.com/issues/43131

**Pull request merged**